relay-pattern/src/lib.rs

```diff
@@ -197,6 +229,88 @@ impl<'a> PatternBuilder<'a> {
     }
 }
 
+/// A builder for a collection of [`Patterns`].
+#[derive(Debug)]
+pub struct PatternsBuilder {
+    options: Options,
+}
+
+impl PatternsBuilder {
+    /// If enabled matches the pattern case insensitive.
+    ///
+    /// This is disabled by default.
+    pub fn case_insensitive(&mut self, enabled: bool) -> &mut Self {
+        self.options.case_insensitive = enabled;
+        self
+    }
+
+    /// Returns a [`PatternsBuilderConfigured`] builder which allows adding patterns.
+    pub fn patterns(&mut self) -> PatternsBuilderConfigured {
+        PatternsBuilderConfigured {
+            strategies: Vec::new(),
+            options: self.options,
+        }
+    }
+
+    /// Adds a pattern to the builder and returns the resulting [`PatternsBuilderConfigured`].
+    pub fn add(&mut self, pattern: &str) -> Result<PatternsBuilderConfigured, Error> {
+        let mut builder = PatternsBuilderConfigured {
+            strategies: Vec::with_capacity(1),
+            options: self.options,
+        };
+        builder.add(pattern)?;
+        Ok(builder)
+    }
+}
+
+/// A [`PatternsBuilder`] with all options configured.
+///
+/// The second step after [`PatternsBuilder`].
+#[derive(Debug)]
+pub struct PatternsBuilderConfigured {
+    strategies: Vec<MatchStrategy>,
+    options: Options,
+}
+
+impl PatternsBuilderConfigured {
+    /// Adds a pattern to the builder.
+    pub fn add(&mut self, pattern: &str) -> Result<&mut Self, Error> {
+        let mut parser = Parser::new(pattern, self.options);
+        parser.parse().map_err(|kind| Error {
+            pattern: pattern.to_owned(),
+            kind,
+        })?;
+
+        let strategy =
+            MatchStrategy::from_tokens(parser.tokens, self.options).map_err(|kind| Error {
+                pattern: pattern.to_owned(),
+                kind,
+            })?;
+
+        self.strategies.push(strategy);
+
+        Ok(self)
+    }
+
+    /// Builds a [`Patterns`] from the contained patterns.
+    pub fn build(self) -> Patterns {
+        Patterns {
+            strategies: self.strategies,
+            options: self.options,
+        }
+    }
+
+    /// Returns [`Patterns`] containing all added patterns and removes them from the builder.
+    ///
+    /// The builder can still be used afterwards, it keeps the configuration.
+    pub fn take(&mut self) -> Patterns {
```

Comment:
I probably need to read the caller side, but why is there a `take()` method in addition to `build()`? If it's about keeping the config, would an additonal `options()` getter method make more sense?

Reply:
It's useful because if the other combinators return `&mut self` you cannot call `build()` because it takes `self`. With `take()` taking `&mut self` you can nicely chain it.